File: algo-c-to-_/src/sierpinski.py
#
#	from src/sierpin.c
#
#	void urd(int)		to	sierpinski; urd
#	void lur(int)		to	sierpinski; lur
#	void dlu(int)		to	sierpinski; dlu
#	void rdl(int)		to	sierpinski; rdl
#

def sierpinski(plotter, order, n):
    def urd(i): # UpRightDown
        if i == 0:
            return
        urd(i - 1)
        plotter.drawRel(h, h)
        lur(i - 1)
        plotter.drawRel(2 * h, 0)
        rdl(i - 1)
        plotter.drawRel(h, -h)
        urd(i - 1)

    def lur(i): # LeftUpRight
        if i == 0:
            return
        lur(i - 1)
        plotter.drawRel(-h, h)
        dlu(i - 1)
        plotter.drawRel(0, 2 * h)
        urd(i - 1)
        plotter.drawRel(h, h)
        lur(i - 1)

    def dlu(i): # DownLeftUp
        if i == 0:
            return
        dlu(i - 1)
        plotter.drawRel(-h, -h)
        rdl(i - 1)
        plotter.drawRel(-2 * h, 0)
        lur(i - 1)
        plotter.drawRel(-h, h)
        dlu(i - 1)

    def rdl(i): # RightDownLeft
        if i == 0:
            return
        rdl(i - 1)
        plotter.drawRel(h, -h)
        urd(i - 1)
        plotter.drawRel(0, -2 * h)
        dlu(i - 1)
        plotter.drawRel(-h, -h)
        rdl(i - 1)

    h, n = 1, n // 6
    for _ in range(1, order):
        h = 3 * h / (6 + h)
    h = h * n

    plotter.move(h + 1, 1)
    urd(order)
    plotter.drawRel(h, h)
    lur(order)
    plotter.drawRel(-h, h)
    dlu(order)
    plotter.drawRel(-h, -h)
    rdl(order)
    plotter.drawRel(h, -h)

File: porting-from-algo-c/examples/sierpinski.lua
--
--  from src/sierpin.c
--
--    void urd(int)                to  sierpinski; urd
--    void lur(int)                to  sierpinski; lur
--    void dlu(int)                to  sierpinski; dlu
--    void rdl(int)                to  sierpinski; rdl
--

local M = require 'svgplot'

local SvgPlotA = M.SvgPlot
local SvgPlotB = M.SvgPlotWholeBuffer
local SvgPlotC = M.SvgPlotWithBuffer
local sierpinski = require 'sierpinski'.sierpinski
local file = require '_helper'.file

local function sampleWriter(pathPrefix, size, offset)
	local m = size + offset
	local pltA, pltB, pltC = SvgPlotA(m, m), SvgPlotB(m, m), SvgPlotC(m, m)

	return function (n)
		function body(plotter)
			plotter:pathStart()
			sierpinski(plotter, n, size)
			plotter:pathEnd(true, M.SV.PRESET_PLAIN)
		end

		file(("%s-A-%d.svg"):format(pathPrefix, n), "w", function (fh)
			pltA:write(fh, body)
		end)

		file(("%s-B-%d.svg"):format(pathPrefix, n), "w", function (fh)
			pltB:write(fh, body):reset()
		end)

		file(("%s-C-%d.svg"):format(pathPrefix, n), "w", function (fh)
			pltC:write(fh, body)
		end)
	end
end

do
	local writer = sampleWriter("results/sierpinski", 600, 2)

	for n=2,6,2 do
		writer(n)
	end
end
